### DISCUSSION TITLE
Arduino IDE Warnings using drawString

### DISCUSSION THREAD

**Discussion by TobbeG** [lovyan03/LovyanGFX#563] [Q&A]
Why do I get warnings using Arduini IDE to compile code for ESP32-S3 using LovyanGFX library and what should I do to fix this - I have no declaration in my code for "drawString(....)" ?

In file included from C:\Users\XXXXXX\Ohmigo.server_3.1.1\Ohmigo.server_3.1.1.ino:464:

**C:\Users\XXXXXX\Dropbox\Ohmigo.server_3.1.1\Display.h:144:114: warning: 'size_t lgfx::v1::LGFXBase::drawString(const String&, int32_t, int32_t, uint8_t)' is deprecated: use IFont [-Wdeprecated-declarations]**

In file included from C:\Users\XXXXXX\Dropbox\MAS\libraries\arduino_188415\src/lgfx/v1_init.hpp:25,
                 from C:\Users\XXXXXX\libraries\arduino_188415\src/LovyanGFX.hpp:31,
                 from C:\Users\XXXXXX\Ohmigo.server_3.1.1\ElecrowSettings.h:2,
                 from C:\Users\XXXXXX\Ohmigo.server_3.1.1\Ohmigo.server_3.1.1.ino:220:
                 
C:\Users\Thorbj rn Geiser O\Dropbox\MAS\libraries\arduino_188415\src/lgfx/v1/LGFXBase.hpp:682:19: note: declared here

     **inline size_t drawString(const String& string, int32_t x, int32_t y, uint8_t      font) { return draw_string(string.c_str(), x, y, _text_style.datum, fontdata[font]); }**
                   ^~~~~~~~~~

Thank you for helping out!

**Comment by tobozo** [comment]
hi, 

> warning: 'size_t lgfx::v1::LGFXBase::drawString(const String&, int32_t, int32_t, uint8_t)' is deprecated: use IFont [-Wdeprecated-declarations]

the deprecation message means you shouldn't use `tft.drawString("blah", x, y, fontNumber)` but `drawString("blah", x, y, font)` where `font` is [one of the listed fonts](https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx/v1/lgfx_fonts.hpp#L269-L461):

```cpp
tft.drawString("blah", x, y, &fonts::FreeMono18pt7b));
```

this is the same as doing this:

```cpp
tft.setFont(&fonts::FreeMono18pt7b);
tft.drawString("blah", x, y);
```

**Reply by TobbeG** [reply]
OK Got it.  

Which font is default?

**Reply by tobozo** [reply]
that would be the first in the list `fonts::Font0`, using `tft.setFont(nullptr)` should have the same effect

**Comment by TobbeG** [comment]
So what shall i replace ???? with if i am using default and swapped foonts only by using settextfont(1) and (2) ?

settextfont(1)   -------------->>> drawString("blah", x, y, ?????? )

settextfont(2)  -------------->>>  drawString("blah", x, y, ?????? )

Thank you for your kind assistant!
Best
Thorbjörn

**Comment by tobozo** [comment]
hi,

settextfont(1)  ~= setFont( &fonts::Font0 );

settextfont(2)  ~= setFont( &fonts::Font2;

**Comment by TobbeG** [comment]
So in my case I just saw I use a few more, but I guess it shall be like this then?

drawString("blah", x, y, 1 )     -->>        drawString("blah", x, y, &fonts::Font1 )
drawString("blah", x, y, 2 )     -->>        drawString("blah", x, y, &fonts::Font2 )
drawString("blah", x, y, 4 )     -->>        drawString("blah", x, y, &fonts::Font4 )
drawString("blah", x, y, 6 )     -->>        drawString("blah", x, y, &fonts::Font6 )

Correct?

**Reply by tobozo** [reply]
yes it would be like this if font numbering made sense, but it doesnt, so don't expect any consistency with the source you got those font numbers from

for example [LovyanGFX's Font0](https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx/v1/lgfx_fonts.cpp#L1182) and [TFT_eSPI's Font1](https://github.com/Bodmer/TFT_eSPI/blob/master/examples/320%20x%20240/Free_Font_Demo/Free_Fonts.h#L258) are both GLCD fonts but with a potentially different size, so trying to match font by index number will lead to mismatches
